src/main/java/com/minecolonies/colony/Colony.java

```diff
@@ -515,110 +452,169 @@ public void updateSubscribers()
                         subscribers.add(player);
                     }
                 }
             }
         }
 
-        //  Determine if any new subscribers were added this pass
-        boolean hasNewSubscribers = false;
-        for (EntityPlayerMP player : subscribers)
-        {
-            if (!oldSubscribers.contains(player))
-            {
-                hasNewSubscribers = true;
-                break;
-            }
-        }
-
         if (!subscribers.isEmpty())
         {
+            //  Determine if any new subscribers were added this pass
+            boolean hasNewSubscribers = hasNewSubscribers(oldSubscribers, subscribers);
+
             //  Send each type of update packet as appropriate:
             //      - To Subscribers if the data changes
             //      - To New Subscribers even if it hasn't changed
 
             //  ColonyView
-            if (isDirty || hasNewSubscribers)
+            sendColonyViewPackets(oldSubscribers, hasNewSubscribers);
+
+            // Permissions
+            sendPermissionsPackets(oldSubscribers, hasNewSubscribers);
+
+            //  Citizens
+            sendCitizenPackets(oldSubscribers, hasNewSubscribers);
+
+            //  Buildings
+            sendBuildingPackets(oldSubscribers, hasNewSubscribers);
+        }
+
+        isDirty = false;
+        isCitizensDirty = false;
+        isBuildingsDirty = false;
+        permissions.clearDirty();
+
+        buildings.values().forEach(AbstractBuilding::clearDirty);
+        citizens.values().forEach(CitizenData::clearDirty);
+    }
+
+    private void sendBuildingPackets(Set<EntityPlayerMP> oldSubscribers, boolean hasNewSubscribers)
+    {
+        if (isBuildingsDirty || hasNewSubscribers)
+        {
+            for (AbstractBuilding building : buildings.values())
             {
-                for (EntityPlayerMP player : subscribers)
+                if (building.isDirty() || hasNewSubscribers)
                 {
-                    boolean isNewSubscriber = !oldSubscribers.contains(player);
-                    if (isDirty || isNewSubscriber)
-                    {
-                        MineColonies.getNetwork().sendTo(new ColonyViewMessage(this, isNewSubscriber), player);
-                    }
+                    ColonyViewBuildingViewMessage msg = new ColonyViewBuildingViewMessage(building);
+
+                    subscribers.stream()
+                               .filter(player -> building.isDirty() || !oldSubscribers.contains(player))
+                               .forEach(player -> MineColonies.getNetwork().sendTo(msg, player));
                 }
             }
+        }
+    }
 
-            // Permissions
-            if(permissions.isDirty() || hasNewSubscribers)
+    private void sendCitizenPackets(Set<EntityPlayerMP> oldSubscribers, boolean hasNewSubscribers)
+    {
+        if (isCitizensDirty || hasNewSubscribers)
+        {
+            for (CitizenData citizen : citizens.values())
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Local variable 'citizen' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)